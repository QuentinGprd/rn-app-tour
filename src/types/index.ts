import { NativeMethods, ViewStyle } from 'react-native';

type highlightRef = {
  current: NativeMethods | null;
};

export type AppTourOptions = {
  /**
   * The element highlighted by the app tour component
   */
  ref: highlightRef;
  /**
   * App tour modal container style
   */
  modalContainerStyle?: ViewStyle;
  /**
   * App tour modal content
   */
  modalComponent?: JSX.Element;
  /**
   * Border radius of the highlighted zone
   * default: 100 (ellipse)
   */
  borderRadiusInPercent: number;
  /**
   * Vertical offset of the highlighted zone
   * Default value: `0`
   */
  yOffset: number;
  /**
   * navigation prop to use for transition animation events (optional)
   */
  navigation?: any;
  /**
   * Called when the app tour background is pressed
   */
  onBackgroundPress?: () => void;
};

export type AppTourMeasure = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export type AppTourUIProps = {
  isVisible: boolean;
  options: AppTourOptions;
};

export type AppTourRef = {
  show: (params: AppTourOptions) => void;
  hide: () => void;
};
