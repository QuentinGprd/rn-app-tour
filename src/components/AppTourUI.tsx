import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { Animated, Dimensions, StyleSheet, View } from 'react-native';
import Svg, { Defs, Mask, Rect, Use } from 'react-native-svg';
import { AppTourMeasure, AppTourUIProps } from '../types';

export const AppTourUI = ({ isVisible, options }: AppTourUIProps) => {
  const {
    ref,
    navigation,
    modalContainerStyle,
    modalComponent,
    borderRadiusInPercent,
    yOffset,
    onBackgroundPress,
  } = options;

  const [measure, setMeasure] = useState<AppTourMeasure>({
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });

  const fadeAnim = useRef(new Animated.Value(0)).current;
  const [animIsFinished, setAnimIsFinished] = useState(false);

  const measureRef = useCallback(
    () =>
      ref?.current?.measureInWindow((x, y, width, height) => {
        if (x && y && width && height) {
          setMeasure({ x, y: y - yOffset, width, height });
        }
      }),
    [ref, yOffset],
  );

  const modalTopPosition = useMemo(() => {
    if (!measure) {
      return 0;
    }

    if (measure.y < Dimensions.get('screen').height / 2) {
      // place modal under mask
      return measure.y + measure.height + 10;
    }

    return 'auto';
  }, [measure]);

  const modalBottomPosition = useMemo(() => {
    if (!measure) {
      return 0;
    }

    if (measure.y >= Dimensions.get('screen').height / 2) {
      // place modal over the mask
      return Dimensions.get('screen').height - measure.y + 10;
    }

    return 'auto';
  }, [measure]);

  const arrowOrientation = modalTopPosition !== 'auto' ? 'top' : 'bottom';

  const arrowLeftOffset = useMemo(() => {
    if (measure) {
      return (
        -Dimensions.get('screen').width / 2 + measure.x + measure.width / 2
      );
    }

    return 0;
  }, [measure]);

  const arrowStyle = useMemo(() => {
    return { ...styles.triangleVertical, left: arrowLeftOffset };
  }, [arrowLeftOffset]);

  useEffect(() => {
    const unsubscribe = navigation?.addListener('transitionEnd', () => {
      if (ref?.current) {
        measureRef();
      }
    });

    return unsubscribe;
  }, [navigation, ref, measureRef]);

  useEffect(() => {
    if (ref?.current) {
      measureRef();
    }
  }, [ref, measureRef]);

  useEffect(() => {
    setAnimIsFinished(false);

    if (measure) {
      Animated.timing(fadeAnim, {
        toValue: isVisible ? 1 : 0,
        duration: 300,
        useNativeDriver: true,
      }).start(({ finished }) => {
        if (!isVisible) {
          setAnimIsFinished(finished);
        }
      });
    }
  }, [fadeAnim, isVisible, measure]);

  if (animIsFinished || !measure) {
    return <></>;
  }

  return (
    <Animated.View
      ref={ref}
      style={{
        opacity: fadeAnim,
        position: 'absolute',
        height: Dimensions.get('screen').height,
        width: Dimensions.get('screen').width,
      }}>
      <Svg
        height={'100%'}
        viewBox={`0 0 ${Dimensions.get('screen').width} ${
          Dimensions.get('screen').height
        }`}
        width={'100%'}
        onPress={onBackgroundPress}>
        <Defs>
          <Mask
            height="100%"
            id="mask"
            /* wait for merge: https://github.com/react-native-svg/react-native-svg/pull/1774
            @ts-ignore */
            maskUnits="userSpaceOnUse"
            width="100%"
            x={0}
            y={0}>
            <Rect fill="white" height="100%" width="100%" />
            <Rect
              fill="black"
              height={measure.height}
              rx={
                measure
                  ? (measure.width * (borderRadiusInPercent / 100)) / 2
                  : 0
              }
              width={measure.width}
              x={measure.x}
              y={measure.y}
            />
          </Mask>
          <Rect
            fill="rgba(0, 0, 0, 0.5)"
            height={'100%'}
            id="bg"
            width={'100%'}
          />
        </Defs>
        <Use fill="none" href="#bg" mask="url(#mask)" />
      </Svg>
      {modalComponent ? (
        <View
          style={{
            bottom: modalBottomPosition,
            left: 10,
            position: 'absolute',
            right: 10,
            top: modalTopPosition,
          }}>
          <View>
            {arrowOrientation === 'top' && <View style={arrowStyle} />}
            <View style={{ flexDirection: 'row' }}>
              <View style={{ ...styles.container, ...modalContainerStyle }}>
                {modalComponent}
              </View>
            </View>
            {arrowOrientation === 'bottom' && (
              <View
                style={{ ...arrowStyle, transform: [{ rotate: '180deg' }] }}
              />
            )}
          </View>
        </View>
      ) : null}
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    borderRadius: 10,
    flex: 1,
    marginTop: -StyleSheet.hairlineWidth,
    paddingHorizontal: 10,
    paddingVertical: 10,
  },
  triangleVertical: {
    alignSelf: 'center',
    width: 0,
    height: 0,
    borderLeftWidth: 10,
    borderRightWidth: 10,
    borderBottomWidth: 10,
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'white',
  },
  triangleHorizontal: {
    alignSelf: 'center',
    width: 0,
    height: 0,
    borderTopWidth: 10,
    borderRightWidth: 10,
    borderBottomWidth: 10,
    borderTopColor: 'transparent',
    borderBottomColor: 'transparent',
    borderRightColor: 'white',
  },
});
