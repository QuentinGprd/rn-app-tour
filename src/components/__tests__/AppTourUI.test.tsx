/* eslint-env jest */

import '@testing-library/jest-native/extend-expect';
import { act, cleanup } from '@testing-library/react-native';
import React, { useRef } from 'react';
import { Text, View } from 'react-native';
import renderer from 'react-test-renderer';
import { AppTourUI } from '../AppTourUI';
import { renderHook } from '@testing-library/react-hooks';
import { AppTourOptions } from '../../types';

let imp: any;

jest.mock('react-native', () => {
  const rn = jest.requireActual('react-native');

  jest
    .spyOn(rn.Animated, 'View', 'get')
    .mockImplementation(() => jest.fn(({ children }) => children));

  imp = jest.spyOn(rn.View.prototype, 'measureInWindow');

  return rn;
});

jest.useRealTimers();
jest.setTimeout(10000);

afterEach(cleanup);

describe('test AppTourUI component', () => {
  it('renders correctly when hidden', async () => {
    const { result: mockedRefHook } = renderHook(() => useRef(null));

    const AppTourModal = (
      <View>
        <Text>App Tour Modal</Text>
      </View>
    );

    const appTourOptions: AppTourOptions = {
      ref: mockedRefHook.current,
      modalComponent: AppTourModal,
      borderRadiusInPercent: 100,
      yOffset: 0,
    };

    let tree;
    await act(async () => {
      tree = renderer.create(
        <View>
          <View ref={mockedRefHook.current}>
            <Text>Hello world!</Text>
          </View>
          <AppTourUI isVisible={false} options={appTourOptions} />
        </View>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with modal on top', async () => {
    imp.mockImplementation(cb => cb(0, 0, 150, 50));

    const { result: mockedRefHook } = renderHook(() => useRef(null));

    const AppTourModal = (
      <View>
        <Text>App Tour Modal</Text>
      </View>
    );

    const appTourOptions: AppTourOptions = {
      ref: mockedRefHook.current,
      modalComponent: AppTourModal,
      borderRadiusInPercent: 100,
      yOffset: 0,
    };

    let tree;
    await act(async () => {
      tree = renderer.create(
        <View>
          <View ref={mockedRefHook.current}>
            <Text>Hello world!</Text>
          </View>
          <AppTourUI isVisible={true} options={appTourOptions} />
        </View>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with modal at bottom', async () => {
    imp.mockImplementation(cb => cb(1000, 1000, 100, 100));

    const { result: mockedRefHook } = renderHook(() => useRef(null));

    const AppTourModal = (
      <View>
        <Text>App Tour Modal</Text>
      </View>
    );

    const appTourOptions: AppTourOptions = {
      ref: mockedRefHook.current,
      modalComponent: AppTourModal,
      borderRadiusInPercent: 100,
      yOffset: 0,
    };

    let tree;
    await act(async () => {
      tree = renderer.create(
        <View>
          <View ref={mockedRefHook.current}>
            <Text>Hello world!</Text>
          </View>
          <AppTourUI isVisible={true} options={appTourOptions} />
        </View>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
