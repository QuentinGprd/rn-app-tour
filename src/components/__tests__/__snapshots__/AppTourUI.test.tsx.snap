// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AppTourUI component renders correctly when hidden 1`] = `
<View>
  <View>
    <Text>
      Hello world!
    </Text>
  </View>
</View>
`;

exports[`test AppTourUI component renders correctly with modal at bottom 1`] = `
<View>
  <View>
    <Text>
      Hello world!
    </Text>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="100%"
    bbWidth="100%"
    focusable={false}
    height="100%"
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": "100%",
          "width": "100%",
        },
      ]
    }
    vbHeight={1334}
    vbWidth={750}
    width="100%"
  >
    <RNSVGGroup>
      <RNSVGDefs>
        <RNSVGMask
          height="100%"
          maskContentUnits={1}
          maskTransform={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
          maskUnits={1}
          name="mask"
          width="100%"
          x={0}
          y={0}
        >
          <RNSVGRect
            fill={4294967295}
            height="100%"
            propList={
              Array [
                "fill",
              ]
            }
            width="100%"
            x={0}
            y={0}
          />
          <RNSVGRect
            fill={4278190080}
            height={100}
            propList={
              Array [
                "fill",
              ]
            }
            rx={50}
            width={100}
            x={1000}
            y={1000}
          />
        </RNSVGMask>
        <RNSVGRect
          fill={2147483648}
          height="100%"
          name="bg"
          propList={
            Array [
              "fill",
            ]
          }
          width="100%"
          x={0}
          y={0}
        />
      </RNSVGDefs>
      <RNSVGUse
        fill={null}
        height={0}
        href="bg"
        mask="mask"
        propList={
          Array [
            "fill",
          ]
        }
        width={0}
        x={0}
        y={0}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      Object {
        "bottom": 344,
        "left": 10,
        "position": "absolute",
        "right": 10,
        "top": "auto",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 10,
              "flex": 1,
              "marginTop": -0.5,
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            }
          }
        >
          <View>
            <Text>
              App Tour Modal
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "borderBottomColor": "white",
            "borderBottomWidth": 10,
            "borderLeftColor": "transparent",
            "borderLeftWidth": 10,
            "borderRightColor": "transparent",
            "borderRightWidth": 10,
            "height": 0,
            "left": 675,
            "transform": Array [
              Object {
                "rotate": "180deg",
              },
            ],
            "width": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`test AppTourUI component renders correctly with modal on top 1`] = `
<View>
  <View>
    <Text>
      Hello world!
    </Text>
  </View>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="100%"
    bbWidth="100%"
    focusable={false}
    height="100%"
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": "100%",
          "width": "100%",
        },
      ]
    }
    vbHeight={1334}
    vbWidth={750}
    width="100%"
  >
    <RNSVGGroup>
      <RNSVGDefs>
        <RNSVGMask
          height="100%"
          maskContentUnits={1}
          maskTransform={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
          maskUnits={1}
          name="mask"
          width="100%"
          x={0}
          y={0}
        >
          <RNSVGRect
            fill={4294967295}
            height="100%"
            propList={
              Array [
                "fill",
              ]
            }
            width="100%"
            x={0}
            y={0}
          />
          <RNSVGRect
            fill={4278190080}
            height={0}
            propList={
              Array [
                "fill",
              ]
            }
            rx={0}
            width={0}
            x={0}
            y={0}
          />
        </RNSVGMask>
        <RNSVGRect
          fill={2147483648}
          height="100%"
          name="bg"
          propList={
            Array [
              "fill",
            ]
          }
          width="100%"
          x={0}
          y={0}
        />
      </RNSVGDefs>
      <RNSVGUse
        fill={null}
        height={0}
        href="bg"
        mask="mask"
        propList={
          Array [
            "fill",
          ]
        }
        width={0}
        x={0}
        y={0}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      Object {
        "bottom": "auto",
        "left": 10,
        "position": "absolute",
        "right": 10,
        "top": 10,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "borderBottomColor": "white",
            "borderBottomWidth": 10,
            "borderLeftColor": "transparent",
            "borderLeftWidth": 10,
            "borderRightColor": "transparent",
            "borderRightWidth": 10,
            "height": 0,
            "left": -375,
            "width": 0,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 10,
              "flex": 1,
              "marginTop": -0.5,
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            }
          }
        >
          <View>
            <Text>
              App Tour Modal
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
